### PR TITLE
Master howl menu dam

### PR DESCRIPTION
**Pull request by Polymorphe57**
This PR has several purposes:
 - add the owl library to Odoo;
 - defines two JavaScript modules web.env and web.test_env that should make
   easier to work with and test owl components.

### CODE REVIEW

File: addons/web/static/src/js/core/owl_dialog.js [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+OwlDialog.defaultProps = {
+    backdrop: 'static',
+    buttons: [],
+    dialogClass: '',
+    fullscreen: false,
+    onForceClose: false,
+    renderFooter: true,
+    renderHeader: true,
+    sizeClass: 'large',
+    subComponent: false,
```
Comment (Polymorphe57) [inline addons/web/static/src/js/core/owl_dialog.js]: remove -> type should always be 'owl.Component' or key absent

File: addons/web/static/src/js/core/owl_dialog.js [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+                    break;
+                }
+            }
+        }
+    }
+
+    willUnmount() {
+        this._removeTooltips();
+        this.$modal.modal('hide');
+        const modals = [...document.getElementsByClassName('modal')];
```
Comment (Polymorphe57) [inline addons/web/static/src/js/core/owl_dialog.js]: On which modal put focus?

File: addons/web/static/src/js/core/owl_dialog.js [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+        this.trigger('dialog_button_clicked', { button });
+    }
+    /**
+     * @private
+     */
+    _onCloseDialog() {
+        this.close();
+    }
+    /**
+     * Manages the TAB key on the buttons. If you the focus is on a primary
```
Comment (Polymorphe57) [inline addons/web/static/src/js/core/owl_dialog.js]: review comment

File: addons/web/static/src/xml/base.xml [hunk 4]
```diff
@@ -5,6 +5,55 @@
 
 <div t-name="Loading" class="o_loading"/>
 
+<t t-name="OwlDialog" owl="1">
+    <div role="dialog"
+        class="modal"
+        t-attf-class="{{props.technical ? 'o_technical_modal' : ''}} {{props.fullscreen ? 'o_modal_full': ''}}"
+        tabindex="-1"
+        data-backdrop="static"
```
Comment (Polymorphe57) [inline addons/web/static/src/xml/base.xml]: we don't use props.backdrop?

File: addons/web/static/tests/core/owl_dialog_tests.js [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+const makeTestEnvironment = require("web.test_env");
+const testUtils = require("web.test_utils");
+
+const targetClassList = [];
+
+let intercepts = {};
+let target = null;
+
+function getTarget(debug) {
```
Comment (Polymorphe57) [inline addons/web/static/tests/core/owl_dialog_tests.js]: should be part of testUtils or something...
Same for other helpers

File: addons/web/static/tests/core/owl_dialog_tests.js [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+    QUnit.test("Closing dialog by clicking on buttons", async function (assert) {
+        assert.expect(2);
+
+        const owlDialog = await createOwlDialog({
+            env: {
+                _t: str => str
+            },
+            props: {
+                onForceClose: () => {
```
Comment (Polymorphe57) [inline addons/web/static/tests/core/owl_dialog_tests.js]: no longer useful

File: addons/web/static/tests/core/owl_dialog_tests.js [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+
+        await testUtils.dom.triggerEvent(document.querySelector('div.modal[role="dialog"]'), 'keydown', { which: 27 });
+        await testPromise;
+
+        assert.verifySteps(['ok custom callback']);
+
+        owlDialog.close();
+    });
+
+    QUnit.skip("Reactivity to props update", async function (assert) {
```
Comment (Polymorphe57) [inline addons/web/static/tests/core/owl_dialog_tests.js]: let  us not forget to rewrite this. the idea was just to test reactivity to updates (button disabled or other things like that)

File: addons/web/static/tests/helpers/test_utils_dom.js [hunk 8]
```diff
@@ -12,6 +12,18 @@ var concurrency = require('web.concurrency');
  * testUtils file.
  */
 
+ /**
+  * Helper function that can wait for the next animation frame according
+  * to the given parameter.
+  *
+  * @param {boolean} nextFrame wait for next animation frame
+  */
+ async function wait(nextFrame) {
```
Comment (Polymorphe57) [inline addons/web/static/tests/helpers/test_utils_dom.js]: should be exported maybe?

File: addons/web/static/src/js/core/owl_dialog.js [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+     */
+    _onButtonClick(buttonIndex) {
+        const button = this.props.buttons[buttonIndex];
+        this.trigger('dialog_button_clicked', { button });
+    }
+    /**
+     * @private
+     */
+    _onCloseDialog() {
+        this.close();
```
Comment (Polymorphe57) [inline addons/web/static/src/js/core/owl_dialog.js]: when is it used?

**Comment by xmo-odoo** [comment]
@Polymorphe57 @Arcasias you'll have to rebase / cherry-pick the branch as a bunch of commits had to be removed from master

File: addons/web/static/tests/helpers/test_env.js [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+        const fakeSession = Object.assign({}, env.session);
+        // TODO: allow to use mockReadGroup,... ?
+
+        if (!templates) {
+            await session.is_bound;
+            templates = session.templatesString
+                .split('t-transition')
+                .join('transition');
+        }
+        const qweb = new OdooQWeb({ translateFn: _t });
```
Comment (Polymorphe57) [inline addons/web/static/tests/helpers/test_env.js]: use test_env_t as translateFn

Comment (Arcasias) [reply addons/web/static/tests/helpers/test_env.js]: Ok

File: addons/web/static/src/js/core/owl_dialog.js (hunk 1 above)
Comment (Arcasias) [reply addons/web/static/src/js/core/owl_dialog.js]: Ok

File: addons/web/static/src/js/core/owl_dialog.js (hunk 2 above)
Comment (Arcasias) [reply addons/web/static/src/js/core/owl_dialog.js]: Ok

File: addons/web/static/src/js/core/owl_dialog.js (hunk 3 above)
Comment (Arcasias) [reply addons/web/static/src/js/core/owl_dialog.js]: Ok

File: addons/web/static/src/js/core/session.js [hunk 11]
```diff
@@ -240,6 +239,7 @@ var Session = core.Class.extend(mixins.EventDispatcherMixin, {
             return $.get(route).then(function (doc) {
                 if (!doc) { return; }
                 qweb.add_template(doc);
+                self.templatesString = doc;
```
[the remark this answers is not in the document]
Comment (Arcasias) [reply addons/web/static/src/js/core/session.js]: Ok

File: addons/web/static/src/js/env.js [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+    const session = require("web.session");
+    const utils = require("web.utils");
+
+    const _t = core._t;
+
+    /**
+     * Sub QWeb class specialized for Odoo needs. This class handles two
+     * additional concerns:
+     * - only load new owl templates
+     * - handle translations.
```
[the remark this answers is not in the document]
Comment (Arcasias) [reply addons/web/static/src/js/env.js]: Ok

File: addons/web/static/src/js/env.js [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+     *
+     * Note that this class always translates sub templates.  It is not currently
```
[the remark this answers is not in the document]
Comment (Arcasias) [reply addons/web/static/src/js/env.js]: Ok

File: addons/web/static/src/js/tools/debug_manager.js [hunk 14]
```diff
@@ -22,6 +22,7 @@ var DebugManager = Widget.extend({
         this._events = null;
         var debug = odoo.debug;
         this.debug_mode = debug;
+        owl.__info__.mode = debug ? "dev" : "prod";
```
[the remark this answers is not in the document]
Comment (Arcasias) [reply addons/web/static/src/js/tools/debug_manager.js]: ok

File: addons/web/static/src/xml/base.xml (hunk 4 above)
Comment (Arcasias) [reply addons/web/static/src/xml/base.xml]: ok

File: addons/web/static/tests/core/owl_dialog_tests.js (hunk 5 above)
Comment (Arcasias) [reply addons/web/static/tests/core/owl_dialog_tests.js]: ok

File: addons/web/static/tests/core/owl_dialog_tests.js [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+
+async function createOwlDialog(params) {
+    intercepts = params.intercepts || {};
+    target = getTarget(!!params.debug);
+
+    let testEnv = await makeTestEnvironment(params.env, params.use);
+    const owlDialog = new OwlDialog(testEnv, params.props || {});
+    await owlDialog.open();
+    // wait for rendering
+    await testUtils.nextTick();
```
[the remark this answers is not in the document]
Comment (Arcasias) [reply addons/web/static/tests/core/owl_dialog_tests.js]: ok

File: addons/web/static/tests/core/owl_dialog_tests.js (hunk 6 above)
Comment (Arcasias) [reply addons/web/static/tests/core/owl_dialog_tests.js]: ok

File: addons/web/static/tests/core/owl_dialog_tests.js [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+
+        await testUtils.dom.triggerEvent(document.querySelector('div.modal[role="dialog"]'), 'keydown', { which: 27 });
+        await testPromise;
+
+        assert.verifySteps(['ok custom callback']);
+
+        owlDialog.close();
+    });
+
+    QUnit.test("Closing dialog with custom buttons without clicking on buttons", async function (assert) {
```
[the remark this answers is not in the document]
Comment (Arcasias) [reply addons/web/static/tests/core/owl_dialog_tests.js]: ok

File: addons/web/static/tests/core/owl_dialog_tests.js [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+        const testPromise = testUtils.makeTestPromiseWithAssert(assert, 'custom callback');
+
+        const owlDialog = await createOwlDialog({
+            env: {
+                _t: str => str
+            },
+            props: {
+                buttons: [
+                    {
+                        text: "Close",
```
[the remark this answers is not in the document]
Comment (Arcasias) [reply addons/web/static/tests/core/owl_dialog_tests.js]: ok

File: addons/web/static/tests/core/owl_dialog_tests.js (hunk 7 above)
Comment (Arcasias) [reply addons/web/static/tests/core/owl_dialog_tests.js]: ok

File: addons/web/static/tests/helpers/test_env.js [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+     *   use that flag.
+     */
+    class OdooQWeb extends owl.QWeb {
+        _addTemplate(name, elem) {
+            if (elem.getAttribute("owl")) {
+                elem.removeAttribute("owl");
+                super._addTemplate(name, elem);
+            }
+        }
+        _compileNode(elem, ctx) {
```
[the remark this answers is not in the document]
Comment (Arcasias) [reply addons/web/static/tests/helpers/test_env.js]: ok

File: addons/web/static/tests/helpers/test_env.js [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+            navigate,
+            blockUI,
+            unblockUI,
+        }, env.services);
+        let fakeSession = Object.assign({}, env.session);
+        // TODO: call performRPC mocked,...
+
+        if (!templates) {
+            await session.is_bound;
+            templates = session.templatesString.getElementsByTagName("templates")[0].outerHTML;
```
[the remark this answers is not in the document]
Comment (Arcasias) [reply addons/web/static/tests/helpers/test_env.js]: ok

File: addons/web/static/tests/helpers/test_utils_dom.js (hunk 8 above)
Comment (Arcasias) [reply addons/web/static/tests/helpers/test_utils_dom.js]: ok

File: addons/web/static/src/js/core/owl_dialog.js (hunk 9 above)
Comment (Arcasias) [reply addons/web/static/src/js/core/owl_dialog.js]: ok

File: addons/web/static/src/js/odoo_qweb.js [hunk 20]
```diff
@@ -0,0 +1,15 @@
+odoo.define("web.OdooQWeb", async function() {
```
Comment (ged-odoo) [inline addons/web/static/src/js/odoo_qweb.js]: remove this

File: addons/web/static/src/xml/base.xml [hunk 21]
```diff
@@ -5,6 +5,71 @@
...
+            <div class="modal-content">
+                <header t-if="props.renderHeader" class="modal-header">
+                    <h4 class="modal-title">
+                        <t t-raw="props.title"/>
+                        <span class="o_subtitle text-muted small"><t t-esc="props.subtitle"/></span>
+                    </h4>
+                    <button type="button" class="close" data-dismiss="modal" aria-label="Close" tabindex="-1">×</button>
+                </header>
+                <main class="modal-body" t-att-class="props.dialogClass">
+                    <t t-if="props.subComponent"
```
Comment (ged-odoo) [inline addons/web/static/src/xml/base.xml]:
```xml
<t t-slot="default"/>
```

File: addons/web/static/src/xml/base.xml [hunk 22]
```diff
@@ -5,6 +5,71 @@
...
+                        t-component="props.subComponent.component"
+                        t-props="props.subComponent.props"
+                    />
+                    <t t-elif="props.textContent" t-esc="props.textContent"/>
+                </main>
+                <footer t-if="props.renderFooter" class="modal-footer" t-ref="modal-footer">
+                    <button t-foreach="buttons" t-as="button"
```
Comment (ged-odoo) [inline addons/web/static/src/xml/base.xml]:
```xml
<t t-slot="buttonzone"/>
```

File: addons/web/static/src/xml/base.xml [hunk 23]
```diff
@@ -5,6 +5,71 @@
...
+                    <button type="button" class="close" data-dismiss="modal" aria-label="Close" tabindex="-1">×</button>
+                </header>
+                <main class="modal-body" t-att-class="props.dialogClass">
+                    <t t-if="props.subComponent"
+                        t-component="props.subComponent.component"
+                        t-props="props.subComponent.props"
+                    />
+                    <t t-elif="props.textContent" t-esc="props.textContent"/>
+                </main>
+                <footer t-if="props.renderFooter" class="modal-footer" t-ref="modal-footer">
```
Comment (ged-odoo) [inline addons/web/static/src/xml/base.xml]: <t t-slot="footer"/>

File: addons/web/static/src/js/env.js [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+        services: {
+            ajaxJsonRPC,
+            blockUI,
+            getCookie,
+            navigate,
+            reloadPage,
+            setCookie,
+            unblockUI,
+        },
+    };
```
Comment (ged-odoo) [inline addons/web/static/src/js/env.js]: some day we want this:
```

main.js

    const env = require('web.env');
    const session = require("web.session");


    owl.QWeb.DIRECTIVE_NAMES.inherit = 1;
    owl.QWeb.DIRECTIVE_NAMES['inherit-mode'] = 1;

    async function startWebClient() {

        await Promise.all([session.is_bound(), loadMenuItems()];
        env.qweb.addTemplates(session.templatesString);
        Component.env = env;

        const webClient = new WebClient();
        await webclient.mount(document.body);
    }

    startWebClient();
```

File: addons/web/static/src/js/components/custom_file_input.js [hunk 25]
```diff
@@ -0,0 +1,88 @@
...
+         *      to upload multiple files at once.
+         */
+        constructor() {
+            super(...arguments);
+
+            this.fileInputRef = useRef('file-input');
+        }
+
+        //--------------------------------------------------------------------------
+        // Private
```
Comment (aab-odoo) [inline addons/web/static/src/js/components/custom_file_input.js]: handlers

File: addons/web/static/src/xml/base.xml [hunk 26]
```diff
@@ -59,6 +59,19 @@
     </Portal>
 </t>
 
+<t t-name="CustomFileInput" owl="1">
+    <span class="o_file_input" aria-atomic="true">
+        <span class="o_file_input_trigger" t-on-click.stop.prevent="_onTriggerClicked">
```
Comment (aab-odoo) [inline addons/web/static/src/xml/base.xml]: don't stop?

File: addons/web/static/tests/components/custom_file_input_tests.js [hunk 27]
```diff
@@ -0,0 +1,103 @@
...
+    const CustomFileInput = require('web.CustomFileInput');
+    const makeTestEnvironment = require('web.test_env');
+    const testUtils = require('web.test_utils');
+
+    const { Component, tags, useState } = owl;
+    const { xml } = tags;
+
+    QUnit.module('Components', {}, function () {
+
+        // This module cannot be tested as thouroughly as we want it to be:
```
Comment (aab-odoo) [inline addons/web/static/tests/components/custom_file_input_tests.js]: thouroughly thoroughly

File: addons/web/static/src/xml/base.xml [hunk 28]
```diff
@@ -59,6 +59,19 @@
...
+            <t t-slot="default"/>
```
Comment (aab-odoo) [inline addons/web/static/src/xml/base.xml]: add default

File: addons/mail/__manifest__.py [hunk 29]
```diff
@@ -59,7 +59,6 @@
         'static/src/xml/thread.xml',
```
[the remark this answers is not in the document]
Comment (Arcasias) [reply addons/mail/__manifest__.py]: Ok

File: addons/web/static/src/js/chrome/action_adapter.js [hunk 30]
```diff
@@ -0,0 +1,124 @@
...
+     * @param {Object} [options={}]
+     * @param {Object} [options.env]
+     */
+    const ActionAdapter = AbstractAction.extend(StandaloneFieldManagerMixin, {
+        /**
+         * TODO: remove this temporary adapter system when the action manager can
+         * handle owl components.
+         * @constructor
+         */
+        init(parent, action, owlAction, options={}) {
```
[the remark this answers is not in the document]
Comment (Arcasias) [reply addons/web/static/src/js/chrome/action_adapter.js]: Ok

File: addons/web/static/src/js/chrome/action_adapter.js [hunk 31]
```diff
@@ -0,0 +1,124 @@
...
+     * @param {Widget} parent @see AbstractAction
+     * @param {Object} action @see AbstractAction
+     * @param {Object} owlAction
+     * @param {Component} owlAction.component the nested owl action class.
+     * @param {Object} owlAction.props the arguments which will be used to instanciate
+     *      the component object.
+     * @param {Object} [options={}]
+     * @param {Object} [options.env]
+     */
+    const ActionAdapter = AbstractAction.extend(StandaloneFieldManagerMixin, {
```
[the remark this answers is not in the document]
Comment (Arcasias) [reply addons/web/static/src/js/chrome/action_adapter.js]: Ok

File: addons/web/static/src/js/chrome/action_adapter.js [hunk 32]
```diff
@@ -0,0 +1,124 @@
...
+
+        /**
+         * Simulate an Owl event coming from the original target as if it were
+         * an actual component.
+         * @private
+         * @param {HTMLElement} target
+         * @param {string} eventType
+         * @param {Object} payload
+         */
+        _trigger(target, eventType, payload) {
```
[the remark this answers is not in the document]
Comment (Arcasias) [reply addons/web/static/src/js/chrome/action_adapter.js]: Ok

File: addons/web/static/src/js/components/custom_file_input.js (hunk 25 above)
Comment (Arcasias) [reply addons/web/static/src/js/components/custom_file_input.js]: Ok

File: addons/web/static/src/xml/base.xml (hunk 26 above)
Comment (Arcasias) [reply addons/web/static/src/xml/base.xml]: Ok

File: addons/web/static/tests/components/custom_file_input_tests.js (hunk 27 above)
Comment (Arcasias) [reply addons/web/static/tests/components/custom_file_input_tests.js]: Ok

**Comment by ged-odoo** [comment]
robodoo r+

**Comment by ged-odoo** [comment]
robodoo rebase-ff

**Pull request closed without merging**